File: api/data/delta/document.go
package delta

const (
	ListTypeUnordered = "UNORDERED"
	ListTypeOrdered   = "ORDERED"
)

type Document struct {
	Operations []*Operation `json:"operations"`
}

// Represents a single operation. Exactly one of Insert, Retain, or Delete should be present.
type Operation struct {
	Insert *InsertOperation `json:"insert,omitempty"`
	Retain *RetainOperation `json:"retain,omitempty"`
	Delete *DeleteOperation `json:"delete,omitempty"`
}

type InsertOperation struct {
	Content    string      `json:"content"`
	Attributes *Attributes `json:"attributes,omitempty"`
}

type RetainOperation struct {
	Length     int         `json:"length"`
	Attributes *Attributes `json:"attributes,omitempty"`
}

type DeleteOperation struct {
	Length int `json:"length"`
}

type RelationshipCardId string

type Attributes struct {
	Emphasis         *bool               `json:"emphasis,omitempty"`
	Strong           *bool               `json:"strong,omitempty"`
	ListType         *string             `json:"listType,omitempty"`
	ListLevel        *int                `json:"listLevel,omitempty"`
	HeadingLevel     *int                `json:"headingLevel,omitempty"`
	RelationshipCard *RelationshipCardId `json:"relationshipCard,omitempty"`
}

func NewDocument() *Document {
	return &Document{
		Operations: []*Operation{},
	}
}

File: web/src/relationship/Document.ts
import { List } from "immutable";

export class Document {
  private operations: Array<DocOperation>;

  constructor() {
    this.operations = [];
  }

  public insert(content: string) {
    this.operations.push(new InsertOperation(content))
  }

  public getOperations(): List<DocOperation> {
    return List.of(...this.operations);
  }
}

export interface DocOperation {
  insert?: string;
}

class InsertOperation implements DocOperation {
  insert?: string

  constructor(content: string) {
    this.insert = content;
  }
}
